smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,152 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.usertune.element;
+
+import java.net.URI;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+
+/**
+ * {@link ExtensionElement} that contains the usertune.
+ */
+public final class UserTuneElement implements ExtensionElement{
+
+    public static final String NAMESPACE = "http://jabber.org/protocol/tune";
+    public static final String ELEMENT = "tune";
+
+    private final String artist;
+    private final int length;
+    private final int rating;
+    private final String source;
+    private final String title;
+    private final int track;
+    private final URI uri;
+
+    private UserTuneElement(UserTuneElementBuilder builder) {
+        this.artist = builder.artist;
+        this.length = builder.length;
+        this.rating = builder.rating;
+        this.source = builder.source;
+        this.title = builder.title;
+        this.track = builder.track;
+        this.uri = builder.uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        if (isNullUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+        String userTuneXML = "<artist>" + artist + "</artist>" +
+                "<length>" + length + "</length>" +
+                "<rating>" + rating + "</rating>" +
+                "<source>" + source + "</source>" +
+                "<title>" + title + "</title>" +
+                "<track>" + track + "</track>" +
+                "<uri>" + uri + "</uri>";
+        return userTuneXML;
+    }
+
+    private boolean isNullUserTune() {
+        if (this.artist == null && this.length == 0 && this.rating == 0 && this.source == null &&
+            this.title == null && this.track == 0 && this.uri == null) {
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    public static boolean hasUserTuneElement(Message message) {
+        return message.hasExtension(ELEMENT,NAMESPACE);
+    }
+
+    public static UserTuneElement fromMessage(Message message) {
+        return message.getExtension(ELEMENT,NAMESPACE);
+    }
+
+    public UserTuneElement getUserTune() {
+        return this;
+    }
+
+    public static class UserTuneElementBuilder{
```

Comment:
Some things about the Builder:  
Usually (but not necessarily) a Builder is simply named `Builder`. In this case it would make sense, as `UserTuneElement.Builder` is specific enough and `UserTuneElement.UserTuneElementBuilder` gets a little bit too long ;)  
Also a Builder has a private constructor and is created via a static method within the class that it is used to build (in this case that would be `public static UserTuneElement.Builder builder()` which belongs into the `UserTuneElement` class.  
All setter methods of a Builder have a return value of the Builder itself (eg. `public Builder setArtist(String artist)`) and return `this`. This comes in handy to chain setter calls, which make the Builder pattern unfold its full potential. Notice how convenient `UserTuneElement.builder().setArtist("Adele").setLength(1234).setRating(9).setXYZ().build()` is ;)  
Lastly, I haven't yet seen the technique to pass a builder instance to the constructor in order to create the object, but I don't see how this is a bad idea either :D

Reply:
Oops, didn't see this before making a commit.
Will act upon this in the following commit.
:)